File: July 2022/July_26.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        List<TreeNode> l1 = helper(root, p), l2 = helper(root, q);
        int i1 = l1.size() - 1, i2 = l2.size() - 1;
        while(i1 >= 0 && i2 >= 0 && l1.get(i1) == l2.get(i2)){
            i1--;
            i2--;
        }
        return l1.get(i1 + 1);
    }
    
    private List<TreeNode> helper(TreeNode root, TreeNode p){
        if(root == null)
            return null;
        if(root == p){
            List<TreeNode> list = new ArrayList<>();
            list.add(p);
            return list;
        }
        
        List<TreeNode> left = helper(root.left, p);
        if(left != null){
            left.add(root);
            return left;
        }
        List<TreeNode> right = helper(root.right, p);
        if(right != null){
            right.add(root);
            return right;
        }
        return null;
    }
}

File: July 2022/July_26.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* root, TreeNode* p, vector<TreeNode*> &v){
        if(root == nullptr)
            return;
        if(root == p){
            v.push_back(p);
            return;
        }
        helper(root->left, p, v);
        if(v.size() > 0){
            v.push_back(root);
            return;
        }
        helper(root->right, p, v);
        if(v.size() > 0){
            v.push_back(root);
            return;
        }
        return;
    }
    
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        vector<TreeNode*> v1, v2;
        helper(root, p, v1);
        helper(root, q, v2);
        int i1 = v1.size() - 1, i2 = v2.size() - 1;
        while(i1 >= 0 && i2 >= 0 && v1[i1] == v2[i2])
            i1--, i2--;
        return v1[i1 + 1];
    }
};
